android/src/main/java/com/tailscale/ipn/DnsConfig.java

```diff
@@ -120,80 +107,47 @@ String getDnsConfigAsString() {
 		return getDnsServersFromNetworkInfo();
 	}
 
-	// getDnsConfigFromLinkProperties finds the DNS servers for each Network interface
-	// returned by ConnectivityManager getAllNetworks().LinkProperties, and return the
-	// one that (heuristically) would be the primary DNS servers.
-	//
-	// on a Nexus 4  with Android  5.1 on wifi: 2602:248:7b4a:ff60::1 10.1.10.1
-	// on a Nexus 7  with Android  6.0 on wifi: 2602:248:7b4a:ff60::1 10.1.10.1
-	// on a Pixel 3a with Android 12.0 on wifi: 2602:248:7b4a:ff60::1 10.1.10.1\nlocaldomain
-	// on a Pixel 3a with Android 12.0 on LTE:  fd00:976a::9 fd00:976a::10
-	//
-	// One odd behavior noted on Pixel3a with Android 12:
-	// With Wi-Fi already connected, starting Tailscale returned DNS servers 2602:248:7b4a:ff60::1 10.1.10.1
-	// Turning off Wi-Fi and connecting LTE returned DNS servers fd00:976a::9 fd00:976a::10.
-	// Turning Wi-Fi back on return DNS servers: 10.1.10.1. The IPv6 DNS server is gone.
-	// This appears to be the ConnectivityManager behavior, not something we are doing.
-	//
-	// This implementation can work through Android 12 (SDK 30). In SDK 31 the
-	// getAllNetworks() method is deprecated and we'll need to implement a
-	// android.net.ConnectivityManager.NetworkCallback instead to monitor
-	// link changes and track which DNS server to use.
-	String getDnsConfigFromLinkProperties() {
-		ConnectivityManager cMgr = (ConnectivityManager) ctx.getSystemService(Context.CONNECTIVITY_SERVICE);
-		if (cMgr == null) {
-			return "";
-		}
-
-		Network[] networks = cMgr.getAllNetworks();
-		if (networks == null) {
-			// Android 6 and before often returns an empty list, but we
-			// can try again with just the active network.
-			//
-			// Once Tailscale is connected, the active network will be Tailscale
-			// which will have 100.100.100.100 for its DNS server. We reject
-			// TYPE_VPN in getPreferabilityForNetwork, so it won't be returned.
-			Network active = cMgr.getActiveNetwork();
-			if (active == null) {
-				return "";
-			}
-			networks = new Network[]{active};
-		}
-
-		// getPreferabilityForNetwork returns an index into dnsConfigs from 0-3.
-		String[] dnsConfigs = new String[]{"", "", "", ""};
-		for (Network network : networks) {
-			int idx = getPreferabilityForNetwork(cMgr, network);
-			if ((idx < 0) || (idx > 3)) {
-				continue;
-			}
-
-			LinkProperties linkProp = cMgr.getLinkProperties(network);
-			NetworkCapabilities nc = cMgr.getNetworkCapabilities(network);
-			List<InetAddress> dnsList = linkProp.getDnsServers();
-			StringBuilder sb = new StringBuilder("");
-			for (InetAddress ip : dnsList) {
-				sb.append(ip.getHostAddress() + " ");
-			}
+	public String getDnsConfigs(){
```

Comment:
I haven't done java in for ever, but I think you could make the locking more idiomatic by using `synchronized` ? Instead of the try/finally blocks and the explicit lock you should be able to write:

```
public String getDnsConfigs() {
  synchronized(this) {
    return this.dnsConfigs;
  }
}
```

And same for the place that updates this.dnsConfigs

Reply:
Done.